shared/AppInsightsCore/src/JavaScriptSDK/BaseTelemetryPlugin.ts

```diff
@@ -77,15 +79,38 @@ export abstract class BaseTelemetryPlugin implements ITelemetryPlugin {
     private _baseTelInit: (config: IConfiguration, core: IAppInsightsCore, extensions: IPlugin[], pluginChain?:ITelemetryPluginChain) => void;
 
     constructor() {
-        let _self = this;
-        let _isinitialized = false;
-        let _rootCtx: IProcessTelemetryContext = null; // Used as the root context, holding the current config and initialized core
-        let _nextPlugin:ITelemetryPlugin|ITelemetryPluginChain = null; // Used for backward compatibility where plugins don't call the main pipeline
+        let _self = this;           // Setting _self here as it's used outside of the dynamicProto as well
+
+        // NOTE!: DON'T set default values here, instead set them in the _initDefaults() function as it is also called during teardown()
+        let _isinitialized: boolean;
+        let _rootCtx: IProcessTelemetryContext; // Used as the root context, holding the current config and initialized core
+        let _nextPlugin: ITelemetryPlugin | ITelemetryPluginChain; // Used for backward compatibility where plugins don't call the main pipeline
+        let _hooks: IInstrumentHook[];
+
+        _initDefaults();
+
+        dynamicProto(BaseTelemetryPlugin, _self, (_self) => {
+
+            _self.initialize = (config: IConfiguration, core: IAppInsightsCore, extensions: IPlugin[], pluginChain?:ITelemetryPluginChain): void => {
+                _setDefaults(config, core, pluginChain);
+                _isinitialized = true;
+            }
+        
+            _self._addHook = (hooks: IInstrumentHook | IInstrumentHook[]) => {
```

Comment:
This will be used to "remember" hooks that have been added for the plugin, so when it's unloaded the hooks will be automatically removed, so each plugin doesn't have to remember them itself.